tests/metrics/timescaledb/test_db.py

```diff
@@ -164,3 +164,28 @@ def test_write(engine, table):
     # check rows are in table
     rows = get_rows(engine, table)
     assert len(rows) == 3
+
+
+def test_upsert(engine, table):
```

Comment:
There's too much logic in this test. When reading it, my attention is focused on what the test function is supposed to be doing, not on what the function under test is supposed to be doing. I don't (yet) have a working dev environment, so I can't add `breakpoint`s to check.